### ISSUE TITLE
Solution to finding a group's closing parenthesis

### ISSUE THREAD

**Issue by hunyan-io** [Lautenschlager-id/expaghetti#1]
A function that accepts a table of chars and returns the position of the closing parenthesis of the first group.
```lua
function findClosing(chars, pos, processing)
    local i = pos or 1
    local increment = processing and 1 or 0
    while i <= #chars do
        while chars[i] == "(" do
            i = findClosing(chars, i+1, true) + increment
        end
        if chars[i] == ")" then
            return i
        end
        i = i + 1
    end
end
```
---
This is an issue that could be detected while trying to parse nested groups, since the childs '(' would be treated as literals and not as new group opening.

**Comment by Lautenschlager-id** [comment]
I'll take a look into your solution to find the best way to bypass this unfortunate problem! Thanks.
It possibly has to be handled somewhere near the last `groupHandler:push(char)` :thinking:

**Comment by tocuto** [comment]
This would fail on something like `([)])`, it would return the first parenthesis, when it should return the second. I'd rather use something like this:

```lua
function findClosing(chars, pos, processing, set)
	local i = pos or 1
	local increment = processing and 1 or 0
	local length = #chars
	if set then
		while i <= length do
			if chars[i] == "]" then
				return i
			end
			i = i + 1
		end
	else
		while i <= length do
			if chars[i] == "[" then
				i = findClosing(chars, i+1, true, true) + increment
			elseif chars[i] == "(" then
				i = findClosing(chars, i+1, true, false) + increment
			end
			if chars[i] == ")" then
				return i
			end
			i = i + 1
		end
	end
end
```

**Comment by Lautenschlager-id** [comment]
I have reached a more reliable version that can be executed during the tree builder.
```Lua
local b = function(str, lim1, lim2)
    local o = { }
    local m = 0
    for i = 1, #str do
        if str[i] == lim1 then
            m = m + 1
            o[m] = { [1] = i }
        elseif str[i] == lim2 then
            o[m][2] = i
            m = m - 1
        end
    end
    return o
end

local str_test = { [0] = "Hello, this (is an expression (nested, maybe?) or not) lol." }
local i = 0
for c in str_test[0]:gmatch('.') do
    i = i + 1
    str_test[i] = c
end
str_test[0] = nil

local result =  b(str_test, '(', ')')
for i = 1, #result do
    print(result[i][1], result[i][2], '=', table.concat(str_test, '', result[i][1], result[i][2]))
end
```

I'll check if it works decently soon and commit the changes.

**Comment by Lautenschlager-id** [comment]
Note:

Successfully implemented on [e81da9f](https://github.com/Lautenschlager-id/expaghetti/commit/e81da9f82d536b674d869b75f897ded2bf61b0d1)
